### DISCUSSION TITLE
Is there any prefetch/preload support (dynamic import)

### DISCUSSION THREAD

**Discussion by lidoravitan** [vercel/next.js#13738] [Help] [answered]
Is there any prefetch/preload support (dynamic import)?
like webpack does [#prefetchingpreloading-modules](https://webpack.js.org/guides/code-splitting/#prefetchingpreloading-modules)

**Comment by jamesmosier** [comment, marked as answer]
Next preloads dynamic components (in production mode) by default. 

From my testing, the other "magic comments" won't work because under the hood this works a bit differently (due to the pre-rendering/SSR capabilities of Next).

Next uses a fork of [react-loadable](https://github.com/jamiebuilds/react-loadable)

**Reply by timneutkens** [reply]
Here's the docs for `next/dynamic`: https://nextjs.org/docs/advanced-features/dynamic-import

**Reply by Emiliano-Bucci** [reply]
@jamesmosier Hi! Technically you could do it by calling `preload()` like in the screen -> 
![image](https://user-images.githubusercontent.com/36669369/106653781-8de51600-6597-11eb-8d6a-7e4e9791491a.png)

@timneutkens Is there a reason why this feature isn't documented? Because i think it's pretty useful; for example, you could preload a js chunk when hovering a button, otherwise if you render for example a modal component that is dynamic imported, when the user click on the button that should show the modal, there would be a bit of delay (because the browser is downloading the js chunk); don't know if i explain myself :D

**Reply by klarstrup** [reply]
For anyone searching this from the beyond: The call signature for this changed as of Next 13.0.7 / https://github.com/vercel/next.js/pull/42589 - it is now `LoadableComponent.preload()` rather than `LoadableComponent.render.preload()`

**Reply by agusterodin** [reply]
As of Next 13.1 it appears that preload method doesn't exist as described above. It doesn't even appear to be an object.

Is there a known fix? 

<img width="639" alt="image" src="https://user-images.githubusercontent.com/10248395/210658981-d3a9a486-c149-46e3-9630-5a094ef581cc.png">

**Reply by klarstrup** [reply]
> As of Next 13.1 it appears that preload method doesn't exist as described above. It doesn't even appear to be an object.
> 
> Is there a known fix?
> 
> <img alt="image" width="639" src="https://user-images.githubusercontent.com/10248395/210658981-d3a9a486-c149-46e3-9630-5a094ef581cc.png">

it's still there as described above 
<img width="316" alt="image" src="https://user-images.githubusercontent.com/1567633/210664100-1839aa3f-21d3-43ad-9837-a0228daa1c78.png">

**Reply by agusterodin** [reply]
I ended up just rendering the component invisibly on the loading overlay screen so that the component is preloaded by the time it is needed 🤮

<img width="392" alt="image" src="https://user-images.githubusercontent.com/10248395/210669032-57939d2a-c607-4016-8803-31ef524d4b8c.png">

**Reply by llllvvuu** [reply]
> Next preloads dynamic components (in production mode) by default.

Where does this happen / when does it happen? On document load?

**Comment by nachaos** [comment]
@timneutkens Just for learning purposes, why the ability to preload all dynamic components is disabled on client side? https://github.com/vercel/next.js/blob/00a14d696dacba19fe4162312ea7aeaaa74ed00f/packages/next-server/lib/loadable.js#L134-L136

Is there a mechanism or workaround to do this? As @Emiliano-Bucci mentioned I think it can be really useful to be able to preload these in order to avoid delays when the browser is downloading the chunk.

**Reply by Emiliano-Bucci** [reply]
@nachaos I've found out that, in addition to the use of the preload method (DynamicComponent.render.preload()), you can also do the following:

```
funcion preloadComponent()	{
   import("component") 
} 
```

By calling the function, the component will be downloaded/preloaded. I guess it's also fine to do it

**Reply by nachaos** [reply]
Thanks for your reply @Emiliano-Bucci, that works as expected. My only concern is that this isn't part of `next/dynamic` public API.

**Reply by sebasusnik** [reply]
> @nachaos I've found out that, in addition to the use of the preload method (DynamicComponent.render.preload()), you can also do the following:
> 
> ```
> funcion preloadComponent()	{
>    import("component") 
> } 
> ```
> 
> By calling the function, the component will be downloaded/preloaded. I guess it's also fine to do it

Hello Emiliano, do you have any example of this? It would be very useful, thanks!

**Comment by teclone** [comment]
I achieved preloading of normal javascript files by using a faked next/dynamic module using this guide:

[How to preload dynamic imports in Next JS](https://www.codecada.com/react/16/improve-critical-rendering-of-tailwindcss-in-nextjs-applications).

the import function imports the desired javascript file but returns a dummy react function. This way, NextJS will treat it as a loadable component

```javascript
import dynamic from `next/dynamic`;

/**
* This will cause NextJS to preload the js file, even though it is not a react component.
* You can also make it a named chunk, using /* webpackChunkName: "name-of-chunk" */
*/
const FakedImport = dynamic(() => {
    return import('../path-to-js-file.js').then(() => {
      // empty react component
      return () => null;
    });
});

export const Component = () => {
  useEffect(() => {
    import('../path-to-js-file.js').then((result) => {
       // do what you want. 
    });
  }, []);
  
  return (
    <>
      <FakedImprt />
      .......
    </>
  );
};
```
